File: sdk/python/matchlock/builder.py
"""Fluent builder for sandbox configuration.

Usage:
    sandbox = Sandbox("python:3.12-alpine") \\
        .allow_host("api.openai.com") \\
        .add_secret("API_KEY", os.environ["API_KEY"], "api.openai.com")

    vm_id = client.launch(sandbox)
"""

from __future__ import annotations

from .types import (
    CreateOptions,
    ImageConfig,
    MountConfig,
    NetworkInterceptionConfig,
    PortForward,
    Secret,
    VFSInterceptionConfig,
)


class Sandbox:
    """Fluent builder for sandbox configuration."""

    def __init__(self, image: str) -> None:
        self._opts = CreateOptions(image=image)

    def with_privileged(self) -> Sandbox:
        self._opts.privileged = True
        return self

    def with_kernel(self, ref: str) -> Sandbox:
        self._opts.kernel_ref = ref
        return self

    def with_cpus(self, cpus: float) -> Sandbox:
        self._opts.cpus = cpus
        return self

    def with_memory(self, mb: int) -> Sandbox:
        self._opts.memory_mb = mb
        return self

    def with_disk_size(self, mb: int) -> Sandbox:
        self._opts.disk_size_mb = mb
        return self

    def with_timeout(self, seconds: int) -> Sandbox:
        self._opts.timeout_seconds = seconds
        return self

    def with_workspace(self, path: str) -> Sandbox:
        self._opts.workspace = path
        return self

    def with_vfs_interception(self, config: VFSInterceptionConfig) -> Sandbox:
        self._opts.vfs_interception = config
        return self

    def with_network_interception(
        self, config: NetworkInterceptionConfig | None = None
    ) -> Sandbox:
        self._opts.force_interception = True
        self._opts.network_interception = config
        return self

    def with_env(self, name: str, value: str) -> Sandbox:
        self._opts.env[name] = value
        return self

    def with_env_map(self, env: dict[str, str]) -> Sandbox:
        self._opts.env.update(env)
        return self

    def allow_host(self, *hosts: str) -> Sandbox:
        self._opts.allowed_hosts.extend(hosts)
        return self

    def block_private_ips(self) -> Sandbox:
        return self.with_block_private_ips(True)

    def with_block_private_ips(self, enabled: bool) -> Sandbox:
        self._opts.block_private_ips = enabled
        self._opts.block_private_ips_set = True
        return self

    def allow_private_ips(self) -> Sandbox:
        return self.with_block_private_ips(False)

    def unset_block_private_ips(self) -> Sandbox:
        self._opts.block_private_ips = False
        self._opts.block_private_ips_set = False
        return self

    def with_network_mtu(self, mtu: int) -> Sandbox:
        self._opts.network_mtu = mtu
        return self

    def with_no_network(self) -> Sandbox:
        self._opts.no_network = True
        return self

    def with_port_forward(self, local_port: int, remote_port: int) -> Sandbox:
        self._opts.port_forwards.append(
            PortForward(local_port=local_port, remote_port=remote_port)
        )
        return self

    def with_port_forward_addresses(self, *addresses: str) -> Sandbox:
        self._opts.port_forward_addresses.extend(addresses)
        return self

    def add_secret(self, name: str, value: str, *hosts: str) -> Sandbox:
        self._opts.secrets.append(Secret(name=name, value=value, hosts=list(hosts)))
        return self

    def add_secret_with_placeholder(
        self, name: str, value: str, placeholder: str, *hosts: str
    ) -> Sandbox:
        self._opts.secrets.append(
            Secret(
                name=name,
                value=value,
                placeholder=placeholder,
                hosts=list(hosts),
            )
        )
        return self

    def with_dns_servers(self, *servers: str) -> Sandbox:
        self._opts.dns_servers.extend(servers)
        return self

    def with_hostname(self, hostname: str) -> Sandbox:
        self._opts.hostname = hostname
        return self

    def mount(self, guest_path: str, config: MountConfig) -> Sandbox:
        self._opts.mounts[guest_path] = config
        return self

    def mount_host_dir(self, guest_path: str, host_path: str) -> Sandbox:
        return self.mount(guest_path, MountConfig(type="host_fs", host_path=host_path))

    def mount_host_dir_readonly(self, guest_path: str, host_path: str) -> Sandbox:
        return self.mount(
            guest_path, MountConfig(type="host_fs", host_path=host_path, readonly=True)
        )

    def mount_memory(self, guest_path: str) -> Sandbox:
        return self.mount(guest_path, MountConfig(type="memory"))

    def mount_overlay(self, guest_path: str, host_path: str) -> Sandbox:
        return self.mount(guest_path, MountConfig(type="overlay", host_path=host_path))

    def with_user(self, user: str) -> Sandbox:
        """Set the user to run commands as (uid, uid:gid, or username)."""
        if self._opts.image_config is None:
            self._opts.image_config = ImageConfig()
        self._opts.image_config.user = user
        return self

    def with_entrypoint(self, *entrypoint: str) -> Sandbox:
        """Override the image entrypoint."""
        if self._opts.image_config is None:
            self._opts.image_config = ImageConfig()
        self._opts.image_config.entrypoint = list(entrypoint)
        return self

    def with_image_config(self, config: ImageConfig) -> Sandbox:
        """Merge the given image configuration into any existing config.

        Fields set in *config* override existing values; falsy fields are left unchanged.
        """
        if self._opts.image_config is None:
            self._opts.image_config = ImageConfig()
        if config.user:
            self._opts.image_config.user = config.user
        if config.working_dir:
            self._opts.image_config.working_dir = config.working_dir
        if config.entrypoint:
            self._opts.image_config.entrypoint = config.entrypoint
        if config.cmd:
            self._opts.image_config.cmd = config.cmd
        if config.env:
            self._opts.image_config.env = config.env
        return self

    def options(self) -> CreateOptions:
        return self._opts

File: sdk/typescript/src/builder.ts
import type {
  CreateOptions,
  ImageConfig,
  MountConfig,
  MountOwnerOptions,
  NetworkInterceptionConfig,
  Secret,
  VFSInterceptionConfig,
} from "./types";

export function cloneCreateOptions(opts: CreateOptions): CreateOptions {
  return {
    ...opts,
    allowedHosts: opts.allowedHosts ? [...opts.allowedHosts] : undefined,
    addHosts: opts.addHosts ? opts.addHosts.map((x) => ({ ...x })) : undefined,
    mounts: opts.mounts
      ? Object.fromEntries(
        Object.entries(opts.mounts).map(([k, v]) => [k, { ...v }]),
      )
      : undefined,
    env: opts.env ? { ...opts.env } : undefined,
    secrets: opts.secrets
      ? opts.secrets.map((s) => ({ ...s, hosts: s.hosts ? [...s.hosts] : undefined }))
      : undefined,
    vfsInterception: opts.vfsInterception
      ? {
        emitEvents: opts.vfsInterception.emitEvents,
        rules: opts.vfsInterception.rules
          ? opts.vfsInterception.rules.map((r) => ({ ...r, ops: r.ops ? [...r.ops] : undefined }))
          : undefined,
      }
      : undefined,
    networkInterception: opts.networkInterception
      ? {
        rules: opts.networkInterception.rules
          ? opts.networkInterception.rules.map((r) => ({
            ...r,
            hosts: r.hosts ? [...r.hosts] : undefined,
            methods: r.methods ? [...r.methods] : undefined,
            setHeaders: r.setHeaders ? { ...r.setHeaders } : undefined,
            deleteHeaders: r.deleteHeaders ? [...r.deleteHeaders] : undefined,
            setQuery: r.setQuery ? { ...r.setQuery } : undefined,
            deleteQuery: r.deleteQuery ? [...r.deleteQuery] : undefined,
            setResponseHeaders: r.setResponseHeaders
              ? { ...r.setResponseHeaders }
              : undefined,
            deleteResponseHeaders: r.deleteResponseHeaders
              ? [...r.deleteResponseHeaders]
              : undefined,
            bodyReplacements: r.bodyReplacements
              ? r.bodyReplacements.map((x) => ({ ...x }))
              : undefined,
          }))
          : undefined,
      }
      : undefined,
    dnsServers: opts.dnsServers ? [...opts.dnsServers] : undefined,
    portForwards: opts.portForwards
      ? opts.portForwards.map((pf) => ({ ...pf }))
      : undefined,
    portForwardAddresses: opts.portForwardAddresses
      ? [...opts.portForwardAddresses]
      : undefined,
    imageConfig: opts.imageConfig
      ? {
        ...opts.imageConfig,
        entrypoint: opts.imageConfig.entrypoint
          ? [...opts.imageConfig.entrypoint]
          : undefined,
        cmd: opts.imageConfig.cmd ? [...opts.imageConfig.cmd] : undefined,
        env: opts.imageConfig.env ? { ...opts.imageConfig.env } : undefined,
      }
      : undefined,
  };
}

export class Sandbox {
  private readonly opts: CreateOptions;

  constructor(image: string) {
    this.opts = { image };
  }

  withPrivileged(): Sandbox {
    this.opts.privileged = true;
    return this;
  }

  withKernel(ref: string): Sandbox {
    this.opts.kernelRef = ref;
    return this;
  }

  withCPUs(cpus: number): Sandbox {
    this.opts.cpus = cpus;
    return this;
  }

  withMemory(mb: number): Sandbox {
    this.opts.memoryMb = mb;
    return this;
  }

  withDiskSize(mb: number): Sandbox {
    this.opts.diskSizeMb = mb;
    return this;
  }

  withTimeout(seconds: number): Sandbox {
    this.opts.timeoutSeconds = seconds;
    return this;
  }

  withWorkspace(path: string): Sandbox {
    this.opts.workspace = path;
    return this;
  }

  withVFSInterception(config: VFSInterceptionConfig): Sandbox {
    this.opts.vfsInterception = config;
    return this;
  }

  withNetworkInterception(config?: NetworkInterceptionConfig): Sandbox {
    this.opts.forceInterception = true;
    this.opts.networkInterception = config;
    return this;
  }

  withEnv(name: string, value: string): Sandbox {
    this.opts.env = this.opts.env ?? {};
    this.opts.env[name] = value;
    return this;
  }

  withEnvMap(env: Record<string, string>): Sandbox {
    this.opts.env = this.opts.env ?? {};
    Object.assign(this.opts.env, env);
    return this;
  }

  allowHost(...hosts: string[]): Sandbox {
    this.opts.allowedHosts = this.opts.allowedHosts ?? [];
    this.opts.allowedHosts.push(...hosts);
    return this;
  }

  addHost(host: string, ip: string): Sandbox {
    this.opts.addHosts = this.opts.addHosts ?? [];
    this.opts.addHosts.push({ host, ip });
    return this;
  }

  blockPrivateIPs(): Sandbox {
    return this.withBlockPrivateIPs(true);
  }

  withBlockPrivateIPs(enabled: boolean): Sandbox {
    this.opts.blockPrivateIPs = enabled;
    this.opts.blockPrivateIPsSet = true;
    return this;
  }

  allowPrivateIPs(): Sandbox {
    return this.withBlockPrivateIPs(false);
  }

  unsetBlockPrivateIPs(): Sandbox {
    this.opts.blockPrivateIPs = false;
    this.opts.blockPrivateIPsSet = false;
    return this;
  }

  addSecret(name: string, value: string, ...hosts: string[]): Sandbox {
    this.opts.secrets = this.opts.secrets ?? [];
    const secret: Secret = { name, value, hosts };
    this.opts.secrets.push(secret);
    return this;
  }

  addSecretWithPlaceholder(
    name: string,
    value: string,
    placeholder: string,
    ...hosts: string[]
  ): Sandbox {
    this.opts.secrets = this.opts.secrets ?? [];
    const secret: Secret = { name, value, placeholder, hosts };
    this.opts.secrets.push(secret);
    return this;
  }

  withDNSServers(...servers: string[]): Sandbox {
    this.opts.dnsServers = this.opts.dnsServers ?? [];
    this.opts.dnsServers.push(...servers);
    return this;
  }

  withHostname(hostname: string): Sandbox {
    this.opts.hostname = hostname;
    return this;
  }

  withNetworkMTU(mtu: number): Sandbox {
    this.opts.networkMtu = mtu;
    return this;
  }

  withNoNetwork(): Sandbox {
    this.opts.noNetwork = true;
    return this;
  }

  withPortForward(localPort: number, remotePort: number): Sandbox {
    this.opts.portForwards = this.opts.portForwards ?? [];
    this.opts.portForwards.push({ localPort, remotePort });
    return this;
  }

  withPortForwardAddresses(...addresses: string[]): Sandbox {
    this.opts.portForwardAddresses = this.opts.portForwardAddresses ?? [];
    this.opts.portForwardAddresses.push(...addresses);
    return this;
  }

  mount(guestPath: string, config: MountConfig): Sandbox {
    this.opts.mounts = this.opts.mounts ?? {};
    this.opts.mounts[guestPath] = { ...config };
    return this;
  }

  mountHostDir(guestPath: string, hostPath: string, opts?: MountOwnerOptions): Sandbox {
    const { ownerUID, ownerGID } = validatedMountOwnerOptions(opts);
    return this.mount(guestPath, { type: "host_fs", hostPath, ownerUID, ownerGID });
  }

  mountHostDirReadonly(guestPath: string, hostPath: string, opts?: MountOwnerOptions): Sandbox {
    const { ownerUID, ownerGID } = validatedMountOwnerOptions(opts);
    return this.mount(guestPath, { type: "host_fs", hostPath, readonly: true, ownerUID, ownerGID });
  }

  mountMemory(guestPath: string): Sandbox {
    return this.mount(guestPath, { type: "memory" });
  }

  mountOverlay(guestPath: string, hostPath: string): Sandbox {
    return this.mount(guestPath, { type: "overlay", hostPath });
  }

  withUser(user: string): Sandbox {
    this.opts.imageConfig = this.opts.imageConfig ?? {};
    this.opts.imageConfig.user = user;
    return this;
  }

  withEntrypoint(...entrypoint: string[]): Sandbox {
    this.opts.imageConfig = this.opts.imageConfig ?? {};
    this.opts.imageConfig.entrypoint = [...entrypoint];
    return this;
  }

  withImageConfig(config: ImageConfig | undefined | null): Sandbox {
    if (!config) {
      return this;
    }
    this.opts.imageConfig = this.opts.imageConfig ?? {};

    if (config.user) {
      this.opts.imageConfig.user = config.user;
    }
    if (config.workingDir) {
      this.opts.imageConfig.workingDir = config.workingDir;
    }
    if (config.entrypoint && config.entrypoint.length > 0) {
      this.opts.imageConfig.entrypoint = [...config.entrypoint];
    }
    if (config.cmd && config.cmd.length > 0) {
      this.opts.imageConfig.cmd = [...config.cmd];
    }
    if (config.env && Object.keys(config.env).length > 0) {
      this.opts.imageConfig.env = { ...config.env };
    }
    return this;
  }

  options(): CreateOptions {
    return cloneCreateOptions(this.opts);
  }
}

export function createSandbox(image: string): Sandbox {
  return new Sandbox(image);
}

function validatedMountOwnerOptions(opts?: MountOwnerOptions): MountOwnerOptions {
  if (!opts) {
    return {};
  }

  validateID("ownerUID", opts.ownerUID)
  validateID("ownerGID", opts.ownerGID)

  return opts;
}

function validateID(name: string, id?: number) {
  if (id !== undefined && (!Number.isInteger(id) || id < 0 || id > 0xffffffff)) {
    throw new RangeError(`${name} must be an integer in [0, 4294967295], got ${id}`);
  }
}
